src/profile-logic/import/simpleperf.js

```diff
@@ -0,0 +1,644 @@
+
+    implementation: [],
+    line: [],
+    column: [],
+
+    length: 0,
+  };
+  frameMap: Map<string, IndexIntoFrameTable> = new Map();
+
+  constructor(strings: UniqueStringArray) {
+    this.strings = strings;
+  }
+
+  toJson(): FrameTable {
+    return this.frameTable;
+  }
+
+  findOrAddFrame(
+    funcIndex: IndexIntoFuncTable,
+    category: IndexIntoCategoryList
+  ): IndexIntoFrameTable {
+    const mapKey = `${funcIndex}-${category}`;
+
+    if (!this.frameMap.has(mapKey)) {
+      this.frameMap.set(mapKey, this.frameTable.length);
+
+      this.frameTable.address.push(-1);
+      this.frameTable.inlineDepth.push(0);
+      this.frameTable.category.push(category);
+      this.frameTable.subcategory.push(0);
+      this.frameTable.func.push(funcIndex);
+      this.frameTable.nativeSymbol.push(null);
+      this.frameTable.innerWindowID.push(null);
+      this.frameTable.implementation.push(null);
+      this.frameTable.line.push(null);
+      this.frameTable.column.push(null);
+
+      this.frameTable.length++;
+    }
+
+    return $NotNull(this.frameMap.get(mapKey));
+  }
+}
+
+class FirefoxSampleTable {
+  strings: UniqueStringArray;
+
+  stackTable: StackTable = {
+    frame: [],
+    category: [],
+    subcategory: [],
+    prefix: [],
+
+    length: 0,
+  };
+  stackMap: Map<string, IndexIntoStackTable> = new Map();
+
+  constructor(strings: UniqueStringArray) {
+    this.strings = strings;
+  }
+
+  toJson(): StackTable {
+    return this.stackTable;
+  }
+
+  findOrAddStack(
+    frameIndex: IndexIntoFrameTable,
+    prefix: IndexIntoStackTable | null,
+    category: IndexIntoCategoryList
+  ): IndexIntoStackTable {
+    const mapKey = `${frameIndex}-${prefix ?? 'null'}`;
+
+    if (!this.stackMap.has(mapKey)) {
+      this.stackMap.set(mapKey, this.stackTable.length);
+
+      this.stackTable.frame.push(frameIndex);
+      this.stackTable.category.push(category);
+      this.stackTable.subcategory.push(0);
+      this.stackTable.prefix.push(prefix);
+
+      this.stackTable.length++;
+    }
+
+    return $NotNull(this.stackMap.get(mapKey));
+  }
+}
+
+class FirefoxThread {
+  name: string;
+  isMainThread: boolean;
+
+  tid: number;
+  pid: number;
+
+  strings = new UniqueStringArray();
+
+  sampleTable: SerializableSamplesTable = {
+    stack: [],
+    time: [],
+    weight: null,
+    weightType: 'samples',
+    length: 0,
+  };
+
+  stackTable: FirefoxSampleTable = new FirefoxSampleTable(this.strings);
+  frameTable: FirefoxFrameTable = new FirefoxFrameTable(this.strings);
+  funcTable: FirefoxFuncTable = new FirefoxFuncTable(this.strings);
+  resourceTable: FirefoxResourceTable = new FirefoxResourceTable(this.strings);
+
+  cpuClockEventId: number = -1;
+
+  constructor(thread: report.IThread) {
+    this.tid = thread.threadId;
+    this.pid = thread.processId;
+
+    this.isMainThread = thread.threadId === thread.processId;
+    this.name = thread.threadName ?? '';
+  }
+
+  toJson(): SerializableThread {
```

Comment:
Any reason you decided to use a `SerializableThread` (and `SerializedProfile` below) instead of a normal `Thread` and (`Profile`)?

Reply:
I migrated this code from somewhere else where I had to use SerializableProfile. Changed now.